### DISCUSSION TITLE
Async Lock

### DISCUSSION THREAD

**Discussion by ackava** [dotnet/csharplang#2611] [Language Ideas]
Currently we cannot write `lock` with `await`.

How about,

```c#
async lock(queue) {
    var next = queue.Dequeue();
    await next.CopyToAsync(dest);
}
```

This will require `AsyncSemaphore` or similar implementation to have `LockAsync` methods to support asynchronous logic.

Implementation, 

```c#
   var s = new SemaphoreLock();
   async using(var l = await s.LockAsync(queue)) {
       // method block here...
   }
```

**Comment by mikedn** [comment]
Normally locks are supposed to be held for short periods of time, otherwise you lose scalability. `async`/`await` tend to imply slowness, otherwise you wouldn't be using `async` in the first place. 

It's not clear from your example why would you need this kind of "async" lock, normally the await would be outside of the lock, unless you really have this rather strange requirement that the queue cannot be accessed until an item already dequeued hasn't finished processing.

**Comment by Joe4evr** [comment]
[`SemaphoreSlim.WaitAsync()`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim.waitasync#System_Threading_SemaphoreSlim_WaitAsync) already exists. And with a small extension method [like this](https://gist.github.com/Joe4evr/55ebac216733a7c21810505380a900af#file-threadingextensions-cs-L31-L35) you can use a regular `using` to guarantee the lock gets released.

**Comment by quinmars** [comment]
Stephen Toub proposed [that solution ](https://devblogs.microsoft.com/pfxteam/building-async-coordination-primitives-part-6-asynclock/) and of course there is Stephen Cleary's [Nito.AsyncEx](https://github.com/StephenCleary/AsyncEx). Also I think such a basic coordinating primitive should be part of netstandard, there is definitely no need for language support.

**Comment by CyrusNajmabadi** [comment]
> > Normally locks are supposed to be held for short periods of time, otherwise you lose scalability. async/await tend to imply slowness, otherwise you wouldn't be using async in the first place.
> > It's not clear from your example why would you need this kind of "async" lock, normally the await would be outside of the lock, unless you really have this rather strange requirement that the queue cannot be accessed until an item already dequeued hasn't finished processing.

> @mikedn
> 
> You don't seem to understand need of async lock here, think of it as non blocking lock same as, Task.Delay is non blocking replacement for Thread.Sleep.
>
> async lock means that thread is free to process some more tasks.

Yes, but that functionality is already *possible* today.  There are existing CoreFx constructs for this.  The language doesn't need to take in any language features here because using these constructs is already *easy* to do.

**Comment by mattwar** [comment]
If I could, I would remove the existing lock statement, because those locks are not clearly defined (too magic) and have the error-prone behavior of allowing reentrancy on the same thread.  Which would be bad for async when you are re-using threads.

**Comment by AartBluestoke** [comment]
async lock is reasonably safe, because a reentrent thread would just hit the same lock and also go await (and become available to complete the continuation of the thread that holds the lock). The lack of visibility over the number of (thread-ish async execution paths-- what do we call these?) concurrently async-lock awaiting could be an issue, but that's the same for all aysnc, not just locks.
